### ISSUE TITLE
acc/unacc gracenotes

### ISSUE THREAD

**Issue by craigsapp** [rism-ch/verovio#403]
It seems that `@grace="acc"` and `@grace="unacc"` renderings are reversed for grace notes:

<img width="310" alt="screen shot 2016-12-29 at 00 10 12" src="https://cloud.githubusercontent.com/assets/3487289/21539365/2eb5bb7c-cd5b-11e6-8ce1-ea4c3cfb3362.png">

`@grace` data:
http://music-encoding.org/documentation/3.0.0/data.GRACE

`acc`:    Time "stolen" from following note.
`unacc`: Time "stolen" from previous note.

The problem could arise due to the abbreviation `acc`.  This is intended to mean "accented", not an acciaccatura.  

`acc`: appogiatura (no slash on single-note grace notes)
`unacc`: acciaccatura (slash on single-note grace notes)

MEI data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="http://music-encoding.org/schema/3.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="3.0.0">
    <meiHead>
        <fileDesc>
            <titleStmt>
                <title />
            </titleStmt>
            <pubStmt>
                <date>2016-12-29 00:07:30</date>
            </pubStmt>
        </fileDesc>
        <encodingDesc>
            <projectDesc>
                <p>Transcoded from Humdrum with Verovio version 0.9.13-dev-591e2fe</p>
            </projectDesc>
        </encodingDesc>
    </meiHead>
    <music>
        <body>
            <mdiv>
                <score>
                    <scoreDef xml:id="scoredef-000000060462560">
                        <staffGrp xml:id="m-000000176290030">
                            <staffDef xml:id="staffdef-000000175268530" clef.shape="G" clef.line="2" meter.count="4" meter.unit="4" n="1" lines="5" />
                        </staffGrp>
                    </scoreDef>
                    <section xml:id="section-000000141390882">
                        <measure xml:id="measure-L3" n="1" right="end">
                            <staff xml:id="staff-L3F1" n="1">
                                <layer xml:id="layer-L3F1" n="1">
                                    <note xml:id="note-L4F1" dur="32" oct="4" pname="g" accid.ges="n" grace="acc" />
                                    <note xml:id="note-L5F1" dur="2" oct="4" pname="c" accid.ges="n">
                                        <verse xml:id="verse-L5F2" n="1">
                                            <syl xml:id="syl-L5F2">acc</syl>
                                        </verse>
                                    </note>
                                    <note xml:id="note-L6F1" dur="32" oct="4" pname="f" accid.ges="n" grace="unacc" />
                                    <note xml:id="note-L7F1" dur="2" oct="4" pname="d" accid.ges="n">
                                        <verse xml:id="verse-L7F2" n="1">
                                            <syl xml:id="syl-L7F2">unacc</syl>
                                        </verse>
                                    </note>
                                </layer>
                            </staff>
                        </measure>
                    </section>
                </score>
            </mdiv>
        </body>
    </music>
</mei>
```

**Comment by lpugin** [comment]
I followed the guidelines http://music-encoding.org/documentation/3.0.0/cmn/#cmnNotesGrace

**Comment by craigsapp** [comment]
Here is a random webpage comparing the two types of grace notes:

http://musictheory4u.com/13.-Acciaccatura-and-Appoggiatura.html

**Comment by pe-ro** [comment]
The example is indeed "incorrect", but see my comments at https://github.com/music-encoding/music-encoding/issues/368.

**Comment by donbyrd** [comment]
I've never been completely comfortable with these terms, so I looked up appogiatura and acciaccatura in the music dictionaries I have at home, viz. the Harvard Dictionary of Music, 2nd and 4th editions (which are completely different) and the New Oxford Companion to Music. None of them supports the idea that a grace note with a slash thru it is an acciaccatura (and at least two say that's a common misconception). They don't completely agree on how an acciaccatura is notated, but it seems that it's either improvised and not notated at all, or it's notated as a normal note, making a plain old triad (for example) look like a tone cluster! So what does the slash indicate? A _short_ appoggiatura, taking less time than a _long_ one, one without the slash. Also, a long appoggiatura does not necessarily take half the time of the following note, though that's very likely the most common situation. There's much more to be said on the subject, but I don't have time to say it :-) .

One of the composers most often associated with acciaccaturas is Domenico Scarlatti. I have a recording of his D-minor Sonata, K. 141, in which the harpsichordist (Piere Hantai) actually _is_ playing tone clusters over and over; it's quite amazing. I don't have the music and I really shouldn't spend any more time on this; maybe someone whose initials are CS would like to dig it up?

**Comment by donbyrd** [comment]
I just invited (via email) Tim Crawford to comment; he undoubtedly knows more about these ornaments than any of us four.

**Comment by craigsapp** [comment]
I excluded pre-1800 music in the other discussion thread :-)
    https://github.com/music-encoding/music-encoding/issues/368

The primary purpose of the slash is to indicate weakness (short, unaccented), so `@grace="unacc"` should show a slash by default and `@grace="acc"` should not.  If the convention of the notation rendering is opposite to that, that is fine, but in that case, `@stem.mod="none|1slash"` should be used to clarify the visual rendering. Or there could be an instruction in the MEI header to define a rendering of `acc` and `unacc` if that is ever necessary.  If the MEI spec is too agnostic, then I will label all of my grace notes `@grace="unknown"` and guarantee their visual aspect with `@stem.mod="none/1slash"`.

> So what does the slash indicate? A short appoggiatura, taking less time than a long one, one without the slash.

Yes, slash implies short (and to some extent unaccented).  What the Italian label is is beside the point as MEI just has `acc`, `unacc` and `unknown`.  However, `acc` implies that it occurs on the beat and takes away time from the note that follows it.  Is this what you are saying a short appoggiatura is doing?

Here is a line of music from Mozart sonata in D major, k311, 3rd movement (manuscript written by Mozart):

http://imslp.nl/imglnks/usimg/e/e7/IMSLP294460-PMLP01842-Mozart_-_Sonata_in_D_Major__K311_-autograph-.pdf
<img width="1159" alt="screen shot 2016-12-30 at 16 29 45" src="https://cloud.githubusercontent.com/assets/3487289/21574818/ea8298da-ceae-11e6-9116-e7eb996ad14d.png">

The first note of the movement is a grace note and it has a slash, but it is not an appoggiatura.  I call this an acciaccatura since it is not an appoggiatura.

Here is the same music from the "Altes Mozart-Ausgabe", 1878:

http://imslp.nl/imglnks/usimg/f/f9/IMSLP56318-PMLP01842-Mozart_Werke_Breitkopf_Serie_20_KV311.pdf
<img width="1149" alt="screen shot 2016-12-30 at 16 27 22" src="https://cloud.githubusercontent.com/assets/3487289/21574850/761c115a-ceaf-11e6-94b3-5ff7f93bdce3.png">

Where the slashes were removed and the flags set to 3.  Who knows why, because it is different from the manuscript and also later editions such as the Neues Mozart-Ausgabe.

Schirmer edition from 1893:
http://ks.imslp.net/files/imglnks/usimg/d/d9/IMSLP00218-Mozart_-_Piano_Sonata__K_311.pdf
<img width="902" alt="screen shot 2016-12-30 at 16 53 39" src="https://cloud.githubusercontent.com/assets/3487289/21574898/c48070ec-ceb0-11e6-8ccf-24c088383a08.png">

Peters' edition 1938:

http://hz.imslp.info/files/imglnks/usimg/f/f8/IMSLP70207-PMLP01842-KV_311.pdf
<img width="885" alt="screen shot 2016-12-30 at 17 00 40" src="https://cloud.githubusercontent.com/assets/3487289/21574928/88d7ab36-ceb1-11e6-89b4-d67bdf7a84d3.png">



This performer plays them "unaccented":
https://www.youtube.com/watch?v=49oiE8Tj1UU

**Comment by donbyrd** [comment]
What the Italian terms are might be beside the point as far as MEI encoding goes, but confusion over them is likely to impede discussion of MEI.

You say of the first note of Mozart K. 311, 3rd mvmt, that "it's a grace note and it has a slash, but it is not an appoggiatura. I call this an acciaccatura since it is not an appoggiatura." Based on the articles in the reference books I cited, I'd say it certainly is not an acciaccatura, and it _is_ an appoggiatura -- specifically a short appoggiatura, which is (at least normally) unaccented, so it's `unacc`. While a long appoggiatura is `acc`.

**Comment by craigsapp** [comment]
> it *is* an appoggiatura

It is definitely *not* an appoggiatura.  Appoggiaturas are supposed to be accented non-harmonic tones.  A grace note an octave below the melody note cannot be described as being dissonant, whether or not it is accented.

Google supports my definitions:

<img width="642" alt="screen shot 2016-12-30 at 17 59 42" src="https://cloud.githubusercontent.com/assets/3487289/21575206/fde81d9a-ceb9-11e6-8ad2-f546953aa0b5.png">

<img width="642" alt="screen shot 2016-12-30 at 18 00 08" src="https://cloud.githubusercontent.com/assets/3487289/21575209/0cf440fc-ceba-11e6-8234-1c0ab9a18a3b.png">

I.e., acciaccaturas steal time from the previous note and are not accented (played on the beat), while appoggiaturas steal time from the following note and are accented (played on the beat since they stole time from the note that was on the beat).  Most  appoggiaturas are now written as regular notes.

The first grace note of mvmt. 3 is played in the example recording https://www.youtube.com/watch?v=49oiE8Tj1UU
as an acciaccatura (short and before the beat, unaccented).

The opening of sonata k311 contains appoggiatura:

<img width="619" alt="screen shot 2016-12-30 at 18 07 12" src="https://cloud.githubusercontent.com/assets/3487289/21575240/fd682fa8-ceba-11e6-8f1b-2fa4545b114f.png">

Mozart wrote them poorly from my point of view :-)  Modern editions usually convert appoggiaturas into regular notes.  Even in the Mozarts Altes-Ausgabe (1878), these were converted into regular notes:

<img width="1009" alt="screen shot 2016-12-30 at 18 10 32" src="https://cloud.githubusercontent.com/assets/3487289/21575251/4c5fb266-cebb-11e6-8338-570337ceb17b.png">

Probably to clarify that they are not acciaccaturas, as this is the modern conceptualization of a grace note.  The editors probably intended the E#5 in m4 as an acciaccatura based on the music from the start of mvmt. 3.  The sample recording plays it as an appoggiatura, but that is probably because she is using a different edition which notates it as an appoggiatura.  Peters edition from 1938:

<img width="849" alt="screen shot 2016-12-30 at 18 16 09" src="https://cloud.githubusercontent.com/assets/3487289/21575278/1438791c-cebc-11e6-9bb8-164fa22b1a51.png">

**Comment by donbyrd** [comment]
Well, Craig, one thing I think we've shown convincingly is that  terminology for types of grace notes in classical-period and earlier music is not as simple and clear-cut as one might expect! I see your point about the first note of the 3rd movement of Mozart K. 311 not being an appoggiatura. But w/r/t it being an acciaccatura, so far we have three well-known music dictionaries on one side and Google on the other. I'd really like to see the New Grove definition of acciaccatura. But in the meantime I'd say the music dictionaries are way ahead, especially since two of them say specifically that the idea that it's a common misconception that a grace note with a slash through it is an acciaccatura. (In the Harvard Dictionary, 2nd ed. -- by the very influential musicologist Willi Apel -- this is in the article on appoggiaturas.)

At the moment I'm leaning strongly towards the view that the first note of the 3rd movement of Mozart K. 311 is neither an acciaccatura nor appoggiatura; it's just a grace note. As for the question of beat and accent, I think there's no doubt that it's before the beat and unaccented. What do you think?

Incidentally, in the Broder/Presser edition of the Mozart sonatas, published (I believe) in the mid-1950's, the the first note of the 3rd movement of K. 311 is written as a small note with no slash and with _two_ flags! I have no idea why Broder did that, but the preface to the volume might say.

**Comment by craigsapp** [comment]
> As for the question of beat and accent, I think there's no doubt that it's before the beat and unaccented. What do you think?

Yes, I would agree with that :-).  At least that is how I would play it...

Wikipedia, another authoritative source,  is more verbose in its definition of acciaccatura, and this would be the sentence that I pay attention to:

https://en.wikipedia.org/wiki/Ornament_(music)#Acciaccatura

> The acciaccatura in the Classical period is usually performed before the beat and the emphasis is on the main note not the grace note. The appoggiatura long or short has the emphasis on the grace note.

and 
>  It [acciaccatura] is written using a grace note (often a quaver, or eighth note), with an oblique stroke through the stem. 

For appoggiatura:

https://en.wikipedia.org/wiki/Ornament_(music)#Appoggiatura
> The appoggiatura is often written as a grace note prefixed to a principal note and printed in small character, usually without the oblique stroke.

**Comment by donbyrd** [comment]
I've seen many excellent articles in Wikipedia, and contributed to a number of them myself. But in terms of being authoritative, I think three well-known serious (i.e., not addressed to amateurs and with articles by musicologists) music dictionaries vastly outweigh Wikipedia plus Google.

**Comment by craigsapp** [comment]
New Grove online:

Acciaccatura:

> In the 19th century, acciaccatura came to mean quick single grace notes, usually a major or minor 2nd above the main note; these were defined as short appoggiaturas in the 18th century by Quantz (1752) and C.P.E. Bach. 

> This type of acciaccatura is generally notated as a small note of semiquaver value before the main note. The custom of writing these semiquavers [16th notes] as quavers [8th notes] with perpendicular slashes through their flags originated in the 18th-century shorthand notation of single semiquavers (especially in Italian sources), whether ornamental or structural notes: it is easier to draw the second flag across the first rather than parallel to it.

Appoggiaturas:

> The standard long and occasional short appoggiaturas are the norm for the Classical period, but the notation and performance of the appoggiatura before 1750 is a more complex and controversial subject.

Mozart and after has a fairly consistent visual syntax for grace notes, and that should be the default rendering if `@stem.mod` is not specified.

> The most common representation has been by a small note [without a slash], but its notated value need not correspond to the length it has when played.

**Comment by donbyrd** [comment]
Interesting. So the experts don't agree on whether using the term _acciaccatura_ for single grace notes notated as 8ths with a slash across the stem is a newer meaning for the term, or simply a mistake! Well, since it isn't necessary to resolve this question for MEI, I'd be happy to drop the subject, having learned a few things in the process :-) . And I won't object to the term being used that way in the future, unless I forget, of course! Thanks, Craig.

**Comment by craigsapp** [comment]
Did I win?  All right!  Words can change meaning over time, even technical ones.  Musical "mode" is even more messy...

Or same words with different meanings at the same time:
     http://www.viralthread.com/english-words-mean-different-things-britain-america

**Comment by pe-ro** [comment]
@craigsapp:  Yes, you win, for some definition of that term.  But, I think what the discussion points out is the terminological minefield that is music notation.  Because of the confusion they tend to cause, I was trying to avoid using the terms "acciaccatura" and "appoggiatura" entirely.  Hence, "acc" and "unacc".  Even these are subjective, though.  It almost doesn't matter what value is in `@grace`.  Any value will suffice to identify the note as a grace note.  Of course, `@stem.mod` is necessary to indicate the presence/lack of a slash on the stem/flag.

**Comment by donbyrd** [comment]
I think Perry is completely right, and actually Craig said almost exactly the same thing in this thread awhile back; I admit I'm the person that made an issue of the terminology! So yes, Craig wins, "for some definition of that term". And, per my last comment, I won't object to the term _acciaccatura_ being used for grace notes with slashes -- though it really seems that short appogiatura is a more precise term, and the authorities aren't in agreement that the meaning of _acciaccatura_ has changed, vs. it being a mistake to use it this way. But this discussion may have been worth the time anyway to emphasize what a "terminological minefield... is music notation". Also to teach me something about historical terminology for notation, though that's just my selfish take :-) .

**Comment by craigsapp** [comment]
> It almost doesn't matter what value is in `@grace`. Any value will suffice to identify the note as a grace note. Of course, `@stem.mod` is necessary to indicate the presence/lack of a slash on the stem/flag.

Yes, I don't care what things are called, but I want a standardized or predictable behavior in notation rendering applications for grace notes (and other things as well, such as undefined accidental states discussed earlier).  The default case should be that `@grace="acc"` does not have a slash unless specified in `@stem.mod`, and `@grace="unacc"` has a slash unless `@stem.mod` says `none`.  If `@stem.mod` is specified, then this would override the default behavior, of course.  Rendering software should use the suggested guidance unless it is customized to a particular repertory's conventions.

If there is no official standard behavior or guidance for software which uses MEI, then this poses a problem for data interoperability between different rendering programs.  If there will only be verovio, then that is not a problem, and you can delegate such behaviors to verovio developers to decide what should happen when an attribute is undefined.  Alternatively, there should be a way of specifying the rendering behavior of logical attributes for a document within the document (which the rendering software can use or choose to ignore).

If I need to explicitly specify `@stem.mod` for every grace note, I can do that, but I don't see why I have to do that (and it currently is not implemented in verovio, and the default behavior in verovio is backwards from what I want it).

**Comment by pe-ro** [comment]
@craigsapp, I understand what you're asking for.  And I'd be happy to modify MEI to help achieve those results *if there were a standard terminology to describe the so-called standard behavior*.  "Acciacatura" and "appoggiatura", as we've come to agree, won't cut it.  And it suspect that there is no convenient way to label grace notes using a single attribute that would work.  That's why there are 2 attributes -- one to say "this is a grace note" and one to say "this note has a slash".  As you point out, this does lead to the problem of having to say on every grace note whether it has a slash or not.  But, since there's no agreed-upon classification mechanism for grace notes that works in all cases, there's easy way around the extra verbosity.  Even if there were a simple one XML attribute solution, it would involve mixing information from two different domains (visual for the slash, gestural for the time aspect) into a single value.  That's not good practice either.

**Comment by pe-ro** [comment]
> The default case should be that `@grace="acc"` does not have a slash unless specified in `@stem.mod`, and `@grace="unacc"` has a slash unless `@stem.mod` says none. If `@stem.mod` is specified, then this would override the default behavior, of course. Rendering software should use the suggested guidance unless it is customized to a particular repertory's conventions.

I don't have a problem with defaults as long as they can be overridden.  So, this is fine with me.

**Comment by craigsapp** [comment]
OK, I'll tell Laurent to do that in verovio now :-)

**Comment by pe-ro** [comment]
Different?  Yes.  Better?  I'm not sure.

`@slashed` would introduce an inconsistency between how slashes are handled on other/regular notes versus grace notes.  Consistency good or not?

On the other hand, using `@slashed` would make a clear distinction between grace note slashes and other kinds.  It would also make it possible to encode a grace note with two kinds of slashes -- one associated with the grace note (usually through the flag, not the stem) and one associated with the stem.

**Comment by lpugin** [comment]
Up to you, then.

**Comment by craigsapp** [comment]
For me, either way is fine.  SCORE uses `@slashed="true|false"` method which I find less opaque than using `@stem.mod="1slash"`.  I would agree with @pe-ro that the current system is more consistent, as the attributes of `<note>` should not be context-dependent on each other (the rendering of the slash will be context-dependent on graciness, which seems weird to me, but that is OK).

If `@slashed` were to be implemented, then it should be added as part of a `<grace>` sub-element for `<note>` rather than on `<note>` itself:

This would produce an unslashed accented grace note:

```xml
<note grace="acc">
```

This would produce a slashed accented grace note:

```xml
<note>
    <grace grace="acc" slashed="true">
</note>
```

**Comment by pe-ro** [comment]
Let's leave that to a future revision. 😄

**Comment by craigsapp** [comment]
OK for me, since I am doing on-demand conversions :-)

**Comment by lpugin** [comment]
What should we do for Verovio now? Just flip the rendering of `acc` and `unacc`?

**Comment by rettinghaus** [comment]
Yes, but take `@stem.mod` into account as @craigsapp proposed: 

> The default case should be that `@grace="acc"` does not have a slash unless specified in `@stem.mod`, and `@grace="unacc"` has a slash unless `@stem.mod` says none. If @stem.mod is specified, then this would override the default behavior, of course. Rendering software should use the suggested guidance unless it is customized to a particular repertory's conventions.

**Comment by craigsapp** [comment]
The accented grace notes are now correct, but the unaccented ones need a slash:

<img width="334" alt="screen shot 2017-06-09 at 2 25 16 pm" src="https://user-images.githubusercontent.com/3487289/26975319-92c62626-4d1f-11e7-83a2-65780072a04f.png">

**Comment by lpugin** [comment]
Fixed with 457def38630a97be25a3501f9a3796354e469c41
